stl/inc/algorithm

```diff
@@ -4710,6 +4710,131 @@ _SampleIt sample(_PopIt _First, _PopIt _Last, _SampleIt _Dest, _Diff _Count,
+                  && indirectly_copyable<_It, _Out> && uniform_random_bit_generator<remove_reference_t<_Urng>>
+        _Out operator()(_It _First, _Se _Last, _Out _Result, iter_difference_t<_It> _Count, _Urng&& _Func) const {
+            _Adl_verify_range(_First, _Last);
+            if (_Count <= 0) {
+                return _Result;
+            }
+
+            _Rng_from_urng<iter_difference_t<_It>, remove_reference_t<_Urng>> _RngFunc(_Func);
+            if constexpr (forward_iterator<_It>) {
+                auto _Pop_size = _RANGES distance(_First, _Last);
+                if (_Count > _Pop_size) {
+                    _Count = _Pop_size;
+                }
+                return _Sample_selection_unchecked(_Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Last)),
+                    _STD move(_Result), _Count, _Pop_size, _RngFunc);
+            } else {
+                return _Sample_reservoir_unchecked(_Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Last)),
+                    _STD move(_Result), _Count, _RngFunc);
+            }
+        }
+
+        template <input_range _Rng, weakly_incrementable _Out, class _Urng>
+           requires (forward_range<_Rng> || random_access_iterator<_Out>)
+                  && indirectly_copyable<iterator_t<_Rng>, _Out>
+                  && uniform_random_bit_generator<remove_reference_t<_Urng>>
+        _Out operator()(_Rng&& _Range, _Out _Result, range_difference_t<_Rng> _Count, _Urng&& _Func) const {
+            if (_Count <= 0) {
+                return _Result;
+            }
+
+
+            _Rng_from_urng<range_difference_t<_Rng>, remove_reference_t<_Urng>> _RngFunc(_Func);
+            if constexpr (forward_range<_Rng>) {
+                auto _Pop_size = _RANGES distance(_Range);
+                if (_Count > _Pop_size) {
+                    _Count = _Pop_size;
+                }
+                return _Sample_selection_unchecked(
+                    _Ubegin(_Range), _Uend(_Range), _STD move(_Result), _Count, _Pop_size, _RngFunc);
+            } else {
+                return _Sample_reservoir_unchecked(
+                    _Ubegin(_Range), _Uend(_Range), _STD move(_Result), _Count, _RngFunc);
+            }
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se, class _Out, class _Rng>
+        _NODISCARD static _Out _Sample_selection_unchecked(_It _First, const _Se _Last, _Out _Result,
+            iter_difference_t<_It> _Count, iter_difference_t<_It> _Pop_size, _Rng& _RngFunc) {
+            // randomly select _Count elements from [_First, _Last) into _Result
+            _STL_INTERNAL_STATIC_ASSERT(forward_iterator<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+            _STL_INTERNAL_STATIC_ASSERT(weakly_incrementable<_Out>);
+            _STL_INTERNAL_STATIC_ASSERT(indirectly_copyable<_It, _Out>);
+
+            for (; _First != _Last; ++_First, (void) --_Pop_size) {

```

Comment:
The termination condition could equivalently be `_Pop_size > 0`, in which case this function doesn't need `_Last` at all. `_Pop_size` should become the second parameter, making this a two-counted-range function, which I think is unique in the STL.

Reply:
Feel free to update the `std` version to agree, and with the `_Count` change as well. Whoever wrote that garbage should be ashamed! (Yes, it was me.)

Reply:
Changed both versions

Reply:
We cannot move the `_Count` part as that is used for the `_Out` iterator aka `_Get_unwrapped_n(_Dest, _Count)`

Reply:
I meant "feel free to replicate the `_Count` change that we made earlier here into the `std` version".